### PR TITLE
Example vars template for stateless playbook

### PR DESCRIPTION
**Pull request by bradenmacdonald**
https://tasks.opencraft.com/browse/OC-2166

### CODE REVIEW

File: openstack-openedx/vars-template.yml [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+  ALLOW_ALL_ADVANCED_COMPONENTS: true
+  AUTH_USE_OPENID: false
+  CERTIFICATES_ENABLED: true
+  CERTIFICATES_HTML_VIEW: true
+  ENABLE_COMBINED_LOGIN_REGISTRATION: true
+  ENABLE_DISCUSSION_SERVICE: true
+  ENABLE_DISCUSSION_HOME_PANEL: true
+  ENABLE_DISCUSSION_EMAIL_DIGEST: true
+  ENABLE_DJANGO_ADMIN_SITE: true
+  ENABLE_INSTRUCTOR_ANALYTICS: true
```
Comment (pomegranited) [inline openstack-openedx/vars-template.yml]: This feature is [marked deprecated](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/feature_flags/feature_flag_index.html).

File: openstack-openedx/vars-template.yml [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+# Set this to a string for the part of your domain, such as ".example.com"
+# that is shared by the LMS and Studio, to enable session sharing between
+# LMS and the Studio.
+#EDXAPP_SESSION_COOKIE_DOMAIN: '.example.com'
+
+################################################################################
+### HTTPS, proxy, and redirect settings
+
+# Set to false if deployed behind a proxy or load balancer.
+NGINX_SET_X_FORWARDED_HEADERS: True
```
Comment (pomegranited) [inline openstack-openedx/vars-template.yml]: This is a little confusing..  Above, we're recommending `EDXAPP_LMS_BASE_SCHEME: 'https'` by default, but also setting `NGINX_ENABLE_SSL: false` below, which makes it look like we're recommending a load balancer, with SSL termination.  So then `NGINX_SET_X_FORWARDED_HEADERS` should be `False`.

If on the other hand, this is for a single server, https deployment, then we should make that clearer, by uncommenting `NGINX_SSL_CERTIFICATE` and `NGINX_SSL_KEY` below, with `"SET ME"` as values like the other required variables.

I'm not sure which situation is better to recommend, but we should pick one and be consistent.

File: openstack-openedx/vars-template.yml [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+# To use an SSL certificate directly on the edxapp VM (instead of on the load balancer):
+#NGINX_ENABLE_SSL: false
+#NGINX_SSL_CERTIFICATE: 'paste certificate here'
+#NGINX_SSL_KEY: 'paste key here'
+
+# Should http:// URLs be redirected to https:// ?
+NGINX_REDIRECT_TO_HTTPS: false
```
Comment (pomegranited) [inline openstack-openedx/vars-template.yml]: If we're assuming that the SSL certificate will exist on the load balancer (as commented above), we can say:
```
NGINX_REDIRECT_TO_HTTPS: true
NGINX_HTTPS_REDIRECT_STRATEGY: "forward_for_proto"
```

File: openstack-openedx/vars-template.yml (hunk 1 above)
Comment (bradenmacdonald) [reply openstack-openedx/vars-template.yml]: Ok, removing.

File: openstack-openedx/vars-template.yml (hunk 2 above)
Comment (bradenmacdonald) [reply openstack-openedx/vars-template.yml]: Ok, I'll go with the load balancer version for now. Good point.

**Comment by antoviaque** [comment]
@bradenmacdonald Instead of a manually crafted file for this, why not refer to the Ocim templates and/or generate it from them? This is going to take work to keep updated, and tested - I'd rather just tap into the work we already do in Ocim.

**Comment by pomegranited** [comment]
@antoviaque I agree that this is a good idea, but I think generating the example from OCIM would be difficult to do initially because of how customisable the OCIM variable templates are..  lots of key variables are programmatically generated, and it might be hard to disentangle them to produce a good starting template.

We could point people to OCIM for a more complete example though, but I think it's nice to provide an example `vars.yml` to start people off.

**Review by pomegranited — APPROVED** [review]

**Comment by antoviaque** [comment]
> I agree that this is a good idea, but I think generating the example from OCIM would be difficult to do initially because of how customisable the OCIM variable templates are.. lots of key variables are programmatically generated, and it might be hard to disentangle them to produce a good starting template.

@pomegranited  Do you have examples of what would be complex exactly? I had in mind to come up with a base set of mock variables/objects, that contain example values, and then just get the templates from https://github.com/open-craft/opencraft/tree/master/instance/templates/instance/ansible concatenated to generate the sample file. These could even be potentially mock objects to get from the tests, or reuse in them, to ensure we keep them in sync.

**Comment by bradenmacdonald** [comment]
@antoviaque I made this file using the vars templates files from Ocim. I considered just referring people to the Ocim templates, but decided that would not work, because:
* Not all the template files in the `ansible` folder are relevant, so simply `cat *.yml` doesn't work
* The template files are django templates and contain conditions like `{% with smtp_relay_settings=appserver.smtp_relay_settings %}{% if smtp_relay_settings %}` which would be confusing because those are not valid ansible syntax. So it's actually necessary to render them via the django template engine (which is not trivial due to missing context), or to manually go through and understand + replace all the django template directives as appropriate. We are in a much better position to do that than people who aren't familiar with Ocim, so I did that myself.
    - If we used a mock context as you suggested, it wouldn't solve this problem cleanly because the `if` statements that evaluate false would be removed from the end result, rather than merely commented out so that the deployer can decide to use them or not.
* The vars files do not include any of the secret key variables, which are extremely important (Ocim doesn't use a template to add those in to the vars). I manually added them in to the template in this PR.
* The Ocim templates assume that a context is available with data about the instance, and don't currently cross-reference variables. This makes the resulting template harder to edit by hand. I went through and combined the variables to reference each other (e.g. creating one master `OPENEDX_RELEASE` variable instead of the 9 separate variables used in the Ocim template)
* The manually created file I put here has more useful comments than the Ocim templates, and a more useful/logical order to the variables

**Comment by antoviaque** [comment]
@bradenmacdonald Cf my previous remark, I think you can go around most of these comments by creating a mock instance in the Ocim tests with meaningful values, and outputing the configuration that Ocim would generate. This way you don't have to redo any of this work, either manually or programatically. It might look a bit less nice that a manually crafted file, but I would prioritize maintenability over presentation on this - it's going to take us a whole lot less efforts and time to maintain, which in turn will help guarantee that we maintain it at all.

**Comment by bradenmacdonald** [comment]
@antoviaque Ok, how do you like this? https://github.com/open-craft/opencraft/pull/204

It's definitely easier to maintain, but since even it has a non-zero maintenance cost and it adds cruft to the codebase, I'd like to suggest that we just post the output of that script for now into a gist, and only add tests and merge that PR if we find that there is community interest.

**Comment by antoviaque** [comment]
@bradenmacdonald Thanks! I've made a few comments on the PR. No worries keeping it on the side for now - although a revert is also fairly easy, so not sure if that would actually change much at the end.

**Pull request closed without merging**